File: 792-number-of-matching-subsequences/792-number-of-matching-subsequences.go
func numMatchingSubseq(s string, words []string) int {
    count := 0
    for _, word := range words {
        if isSubseq(s, word) {
            count += 1
        }
    }
    return count
}

func isSubseq(s, t string) bool {
    j := 0
    for i := 0; i < len(s) && j < len(t); i++ {
        if s[i] == t[j] {
            j += 1
        }
    }
    return j == len(t)
}

File: 792-number-of-matching-subsequences/792-number-of-matching-subsequences.py
class Solution:
    def numMatchingSubseq(self, s: str, words: List[str]) -> int:
        word_dict = defaultdict(list)
        count = 0
        
        for w in words:
            word_dict[w[0]].append(w)
        
        for ch in s:
            if ch not in word_dict:
                continue
            for w in word_dict.pop(ch):
                if len(w) == 1:
                    count += 1
                else:
                    word_dict[w[1]].append(w[1:])
        
        return count